### PR TITLE
feat: creating an SMT verification module

### PR DESCRIPTION
**Pull request by Sarkoxed**
# Description

Please provide a paragraph or two giving a summary of the change, including relevant motivation and context.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] The branch has been merged with/rebased against the head of its merge target.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.
- [ ] No superfluous `include` directives have been added.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to any issue(s) it resolves.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.

### CODE REVIEW

File: cpp/src/barretenberg/proof_system/circuit_builder/circuit_builder_base.hpp [hunk 1]
```diff
@@ -187,6 +193,99 @@ template <typename Arithmetization> class CircuitBuilderBase {
         real_variable_tags.emplace_back(DUMMY_TAG);
         return index;
     }
+
+    /**
+     * Assign a name to a variable(equivalence calss). Should be one name per equivalence class.
```
Comment (Rumata888) [inline cpp/src/barretenberg/proof_system/circuit_builder/circuit_builder_base.hpp]: Typo ("Calss")

File: cpp/src/barretenberg/proof_system/circuit_builder/standard_circuit_builder.cpp [hunk 2]
```diff
@@ -507,6 +510,48 @@ template <typename FF> bool StandardCircuitBuilder_<FF>::check_circuit()
...
+            this->real_variable_index[w_o[i]],
+        };
+        cir.selectors.push_back(tmp_sel);
+        cir.wits.push_back(tmp_wit);
+    }
+
+    msgpack::sbuffer buffer;
+    msgpack::pack(buffer, cir);
+    return buffer;
+    // info("Buffer size: ", buffer.size());
```
Comment (Rumata888) [inline cpp/src/barretenberg/proof_system/circuit_builder/standard_circuit_builder.cpp]: Commented out code

File: cpp/src/barretenberg/smt_verification/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+barretenberg_module(smt_verification common ${CMAKE_SOURCE_DIR}/src/cvc5/tmp-lib/lib/libcvc5.so.1)
+
+barretenberg_module(smt_examples common proof_system stdlib_primitives smt_verification)
+#barretenberg_module(smt_polynomials common proof_system stdlib_primitives smt_verification)
```
Comment (Rumata888) [inline cpp/src/barretenberg/smt_verification/CMakeLists.txt]: Commented out code

File: cpp/src/barretenberg/smt_verification/README.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+- ```export_circuit()``` - exports all variables, gates, and assigned names to an msgpack-compatible buffer namely `msgpack::sbuffer`.
+
+- ```export_circuit_json()``` - exports all variables, gates, and assigned names to a json(kinda)-compatible string.
+
+## 2. Symbolic Circuit initialization and term creation
+
+1. First you need to import the circuit from the saved file or from the buffer:
+
+	- `smt_circuit::CircuitSchema c_info = smt_circuit::unpack_from_file(str fname);`
```
Comment (Rumata888) [inline cpp/src/barretenberg/smt_verification/README.md:51]: Can you provide instructions without storing the circuit on disk?

File: cpp/src/barretenberg/smt_verification/check.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+# TODO how exactly ref works
+import cvc5
```
Comment (Rumata888) [inline cpp/src/barretenberg/smt_verification/check.py]: This file is not documented at all

File: cpp/src/barretenberg/smt_verification/circuit/circuit.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
+#include "circuit.hpp"
+namespace smt_circuit {
+
+Circuit::Circuit(CircuitSchema& circuit_info, Solver* solver, const std::string& tag)
```
Comment (Rumata888) [inline cpp/src/barretenberg/smt_verification/circuit/circuit.cpp:11]: This file has very few comments

**Comment by codygunton** [comment]
This became a monorepo PR

**Pull request closed without merging**